# AstroWave: del dato astronómico al sonido

## Principio del proyecto

AstroWave se plantea como un proyecto de **sonificación de datos astronómicos**.

La sonificación consiste en utilizar sonido para representar información
contenida en datos. En este contexto, una variable astronómica no se convierte
directamente en sonido: necesitamos definir una regla que determine cómo será
representada mediante uno o más parámetros acústicos.

Por ejemplo:

$$
T \longrightarrow g(T) \longrightarrow f_{\mathrm{sound}}
$$

donde:

- $T$ es la temperatura de una estrella;
- $g(\cdot)$ es una transformación que nosotros diseñamos;
- $f_{\mathrm{sound}}$ es una frecuencia sonora utilizada para representar
  esa temperatura.

La temperatura **no es** una frecuencia acústica. La función $g$ constituye
una decisión de modelación.

---

## Sonificación versus creación sonora inspirada en datos

Es útil distinguir dos enfoques posibles.

### Sonificación de datos

En una sonificación, el sonido busca representar alguna propiedad o relación
presente en los datos.

Por ejemplo, si dos estrellas tienen temperaturas similares, podríamos
esperar que su representación sonora también conserve algún tipo de
proximidad perceptible.

El objetivo principal no es necesariamente que el resultado sea musicalmente
agradable, sino que exista una relación explícita y reproducible entre los
datos y el sonido.

### Creación sonora inspirada en datos

También es posible utilizar datos astronómicos como inspiración para una
composición, una textura sonora o una experiencia artística.

En ese caso, las decisiones pueden responder principalmente a criterios
estéticos y no es necesario que las relaciones entre los datos originales
se conserven en el sonido.

Ambos enfoques son válidos, pero responden a preguntas diferentes.

**AstroWave utilizará la sonificación como principio rector**, aunque las
decisiones de diseño sonoro también podrán considerar aspectos perceptuales
y estéticos.

---

## ¿Qué significa modelar en AstroWave?

Modelar no significa necesariamente construir una ecuación complicada ni
utilizar aprendizaje de máquina.

En este proyecto, modelar comienza simplemente al definir una transformación:

$$
y = g(x)
$$

donde una variable astronómica $x$ se transforma en un parámetro acústico $y$.

Por ejemplo:

$$
f_{\mathrm{sound}} = g(T)
$$

$$
\tau_{\mathrm{delay}} = h(d)
$$

$$
A = q(L)
$$

donde $T$, $d$ y $L$ podrían representar temperatura, distancia y
luminosidad.

**Cada función constituye una hipótesis de representación.**

Por eso, antes de elegir una transformación deberíamos preguntarnos:

- ¿qué propiedad de los datos queremos conservar?
- ¿qué rango tienen los valores originales?
- ¿la relación debe conservar el orden?
- ¿valores astronómicos próximos deberían producir sonidos próximos?
- ¿el parámetro acústico resultante se encuentra en un rango útil o perceptible?
- ¿qué ocurre con valores extremos?

Estas preguntas forman parte del proceso de modelación.

---

## Pipeline conceptual

El proyecto puede entenderse inicialmente como:

$$
\text{datos astronómicos}
\longrightarrow
\text{exploración}
\longrightarrow
\text{modelo de mapeo}
\longrightarrow
\text{parámetros acústicos}
\longrightarrow
\text{sonido}
$$

Solo después de comprender este proceso será necesario decidir cómo presentar
la sonificación mediante una interfaz.

---

## Principios de trabajo

Durante el desarrollo de AstroWave intentaremos mantener cuatro principios:

1. **Datos reales**  
   Las variables utilizadas deben provenir de fuentes astronómicas
   identificables y reproducibles.

2. **Mapeos explícitos**  
   Toda transformación entre una variable astronómica y un parámetro acústico
   debe poder escribirse y explicarse.

3. **Relaciones interpretables**  
   Cuando sea posible, la sonificación debería conservar alguna relación
   relevante presente en los datos originales.

4. **Reproducibilidad**  
   Dados los mismos datos y las mismas reglas de mapeo, el procedimiento
   debe producir el mismo resultado.

Estos principios no determinan cómo debe sonar AstroWave. Su propósito es
ayudarnos a justificar las decisiones que iremos tomando.

---

**El sonido no es el dato: es una representación del dato.**


# Hito 1: Exploración de datos astronómicos

## Antes de sonificar, necesitamos conocer nuestros datos

AstroWave busca transformar variables astronómicas reales en parámetros
acústicos.

Antes de decidir cómo debe sonar una estrella, necesitamos responder preguntas
más básicas:

- ¿de dónde provienen los datos?
- ¿qué variables tenemos disponibles?
- ¿qué unidades utilizan?
- ¿qué rangos presentan?
- ¿existen valores faltantes?
- ¿qué variables podrían ser útiles para una sonificación?

La idea central será mantener siempre explícita la transformación

$$
x \longrightarrow g(x) \longrightarrow y
$$

pero en este hito todavía no definiremos $g$. Primero necesitamos conocer la
estructura, las unidades, los rangos y la calidad de los datos disponibles.

En este primer hito **no generaremos sonido todavía**. Primero exploraremos
datos astronómicos reales y comprenderemos su estructura.


## 1. Primer contacto con una fuente astronómica real

Antes de construir un conjunto de datos o seleccionar variables para la
sonificación, comenzaremos consultando un solo objeto astronómico.

El propósito de esta primera consulta no es todavía construir el conjunto de
datos definitivo, sino aprender a:

- consultar una fuente astronómica desde Python;
- reconocer qué información devuelve;
- identificar unidades y tipos de variables;
- distinguir entre los datos que deseamos utilizar y los que la fuente entrega
  efectivamente.

Comenzaremos con **Betelgeuse**, uno de los objetos utilizados en la
presentación inicial de AstroWave.


In [228]:
import astropy
import astroquery
import pandas as pd
import matplotlib.pyplot as plt

print(f"Astropy:    {astropy.__version__}")
print(f"Astroquery: {astroquery.__version__}")
print(f"Pandas:     {pd.__version__}")

Astropy:    6.1.7
Astroquery: 0.4.11
Pandas:     2.3.3


### Betelgeuse antes de ser una fila de datos

Antes de consultar sus variables astronómicas, observemos el objeto que estamos
comenzando a estudiar.

Betelgeuse (Alpha Orionis) es una estrella supergigante roja. La composición
siguiente, basada en observaciones del **Hubble Space Telescope**, nos recuerda
que las variables que consultaremos —posición, distancia, temperatura, tipo
espectral o luminosidad— describen un objeto físico real.

![Betelgeuse observada por Hubble](images/betelgeuse-location-and-size.webp)

*Composición de Betelgeuse y su ubicación en Orión, con referencias de escala
respecto de las órbitas de la Tierra y Júpiter. Crédito: A. Dupree (CfA),
NASA, ESA / Hubble Space Telescope.*

> **Pregunta para pensar:**  
> Si queremos representar Betelgeuse mediante sonido, ¿qué propiedades físicas
> del objeto querríamos conservar o hacer perceptibles en la sonificación?


In [229]:
from astroquery.simbad import Simbad

result = Simbad.query_object("Betelgeuse")

result

main_id,ra,dec,coo_err_maj,coo_err_min,coo_err_angle,coo_wavelength,coo_bibcode,matched_id
,deg,deg,mas,mas,deg,,,
object,float64,float64,float32,float32,int16,str1,object,object
* alf Ori,88.79293899077537,7.407063995272694,9.040423,5.716997,90,O,2007A&A...474..653V,NAME Betelgeuse


### ¿Qué aprendimos de la primera consulta?

La consulta por nombre encontró correctamente a Betelgeuse, identificada en
SIMBAD como `* alf Ori` (Alpha Orionis), y devolvió una fila con información
básica de identificación, posición y metadatos asociados a esas coordenadas.

Sin embargo, las variables que queremos estudiar para AstroWave —por ejemplo,
temperatura, distancia o tipo espectral— no aparecen automáticamente en esta
consulta básica.

Esto introduce una distinción importante:

$$
\text{variable que queremos utilizar}
\neq
\text{variable disponible automáticamente}
$$

Antes de construir nuestro conjunto de datos debemos conocer qué información
ofrece cada fuente, en qué unidades la entrega y qué campos necesitamos
solicitar explícitamente.

Este es también un principio de reproducibilidad: **no basta con conocer un
valor; necesitamos saber de qué fuente proviene y cómo fue obtenido dentro de
nuestro flujo de trabajo.**


## 2. Solicitar explícitamente nuevas propiedades

La consulta básica nos permitió identificar correctamente a Betelgeuse, pero no
entregó automáticamente todas las variables que podrían interesarnos para una
sonificación.

Ahora ampliaremos la consulta solicitando explícitamente algunas propiedades:

- **tipo de objeto**;
- **tipo espectral**;
- **paralaje**.

Este paso ilustra una idea importante en ciencia de datos:

$$
\text{pregunta del proyecto}
\longrightarrow
\text{variables necesarias}
\longrightarrow
\text{campos disponibles en la fuente}
$$

No debemos asumir que una base de datos contiene exactamente las variables que
imaginamos al formular el proyecto.

In [230]:
from astroquery.simbad import Simbad

simbad = Simbad()

simbad.add_votable_fields(
    "otype",
    "sp_type",
    "parallax",
)

betelgeuse_extended = simbad.query_object("Betelgeuse")

betelgeuse_extended

main_id,ra,dec,coo_err_maj,coo_err_min,coo_err_angle,coo_wavelength,coo_bibcode,otype,sp_type,plx_err_prec,plx_value,plx_qual,plx_bibcode,plx_err,plx_prec,matched_id
,deg,deg,mas,mas,deg,,,,,,mas,,,mas,,
object,float64,float64,float32,float32,int16,str1,object,object,object,int16,float64,str1,object,float32,int16,object
* alf Ori,88.79293899077537,7.407063995272694,9.040423,5.716997,90,O,2007A&A...474..653V,s*r,M1-M2Ia-Iab,2,6.55,A,2007A&A...474..653V,0.83,2,NAME Betelgeuse


### Cómo leer esta tabla sin perdernos

La consulta extendida contiene muchas columnas, pero **no todas cumplen el mismo
papel**. Para comenzar, podemos organizarlas en cuatro familias.

#### 1. Identidad: ¿qué objeto estamos mirando?

- `main_id`: identificador principal utilizado por SIMBAD.
- `matched_id`: nombre que coincidió con nuestra búsqueda.

En este caso, `* alf Ori` y `NAME Betelgeuse` se refieren al mismo objeto:
**Betelgeuse = Alpha Orionis**.

#### 2. Posición: ¿dónde está en el cielo?

- `ra`: ascensión recta.
- `dec`: declinación.

Juntas permiten ubicar el objeto sobre la esfera celeste. Los campos
`coo_err_maj`, `coo_err_min` y `coo_err_angle` describen la incertidumbre de
esa posición; `coo_wavelength` indica la región espectral de origen de las
coordenadas y `coo_bibcode` identifica su referencia bibliográfica.

No necesitamos sonificar estos metadatos, pero nos recuerdan algo importante:

> un dato científico no es solo un número; también tiene **incertidumbre,
> procedencia y contexto de medición**.

#### 3. Clasificación: ¿qué clase de objeto es?

- `otype`: tipo de objeto astronómico.
- `sp_type`: tipo espectral.

Para Betelgeuse aparecen:

```text
otype   = s*r
sp_type = M1-M2Ia-Iab
```

SIMBAD utiliza `s*r` como código compacto para una **supergigante roja**.

El tipo espectral `M1-M2Ia-Iab` puede leerse, de manera simplificada, así:

```text
M1-M2    |    Ia-Iab
   ↓             ↓
tipo y       clase de
subtipo      luminosidad
espectral    (supergigante)
```

- **M** indica una estrella de clase espectral M, asociada a temperaturas
  relativamente bajas y apariencia rojiza;
- **1–2** indica un subtipo dentro de la clase M;
- **Ia–Iab** corresponde a clases de luminosidad de estrellas supergigantes
  muy luminosas.

No necesitamos memorizar esta nomenclatura. Lo importante para AstroWave es
reconocer que `sp_type` **contiene información física** y, por tanto, podría
llegar a participar en una decisión de sonificación.

#### 4. Paralaje: una medición relacionada con la distancia

Los campos principales son:

- `plx_value`: valor medido de la paralaje;
- `plx_err`: incertidumbre asociada;
- `plx_qual`: indicador de calidad;
- `plx_bibcode`: referencia bibliográfica de la medición.

Para Betelgeuse:

$$
\pi = 6.55\ \mathrm{mas},
\qquad
\sigma_\pi = 0.83\ \mathrm{mas}
$$

donde `mas` significa **milisegundos de arco**.

La paralaje es un pequeño desplazamiento angular aparente que puede utilizarse
para obtener información sobre la distancia a una estrella. Por ahora **no la
convertiremos todavía en distancia**: primero queremos comprender qué mide y
qué incertidumbre posee.

Los campos `plx_prec` y `plx_err_prec` son metadatos técnicos sobre la
precisión con que se representan esos valores en la tabla; no son nuevas
propiedades físicas de Betelgeuse.

---

#### ¿Qué nos interesa para AstroWave?

Esta tabla ya permite distinguir entre:

**Propiedades o mediciones que podrían contribuir a una representación sonora**
```text
tipo espectral
paralaje / distancia
(y más adelante: temperatura, luminosidad...)
```

y **metadatos necesarios para comprender y auditar los datos**
```text
incertidumbre
calidad
referencia bibliográfica
identificadores
```

**No todas las columnas tienen que convertirse en sonido.** Elegir qué
variables representar, y cómo hacerlo, forma parte de la modelación del
proyecto.

<details>
<summary><strong>Para profundizar: documentación de SIMBAD</strong></summary>

- [SIMBAD: tipos de objetos](https://simbad.cds.unistra.fr/Pages/guide/otypes.htx)
- [SIMBAD: codificación del tipo espectral](https://simbad.cds.unistra.fr/Pages/guide/chD.htx)
- [Astroquery/SIMBAD: selección y descripción de columnas](https://astroquery.readthedocs.io/en/stable/simbad/simbad.html)

</details>


In [231]:
# Una vista reducida nos ayuda a concentrarnos en las columnas
# más útiles para interpretar este primer ejemplo.

columnas_clave = [
    "main_id",
    "otype",
    "sp_type",
    "plx_value",
    "plx_err",
    "plx_qual",
    "plx_bibcode",
    "matched_id",
]

betelgeuse_resumen = betelgeuse_extended[columnas_clave]

betelgeuse_resumen


main_id,otype,sp_type,plx_value,plx_err,plx_qual,plx_bibcode,matched_id
,,,mas,mas,,,
object,object,object,float64,float32,str1,object,object
* alf Ori,s*r,M1-M2Ia-Iab,6.55,0.83,A,2007A&A...474..653V,NAME Betelgeuse


### Una primera lectura de Betelgeuse

Con esta vista reducida podemos describir nuestro primer objeto sin utilizar
todavía ninguna transformación sonora:

- SIMBAD identifica a Betelgeuse como **Alpha Orionis** (`* alf Ori`);
- está clasificada como una **supergigante roja**;
- su tipo espectral registrado es `M1-M2Ia-Iab`;
- la paralaje mostrada por esta consulta es

$$
\pi = 6.55 \pm 0.83\ \mathrm{mas}.
$$

Todavía no hemos decidido cuál de estas variables debe convertirse en sonido.

Ese será justamente uno de los problemas de modelación de AstroWave:

$$
\text{propiedad física}
\longrightarrow
\text{¿qué queremos representar?}
\longrightarrow
\text{parámetro acústico}
$$

Antes de elegir el mapeo necesitamos seguir identificando qué variables
físicas están disponibles y cuáles son realmente relevantes para el proyecto.

### Pausa antes de seguir

Hasta aquí todavía **no hemos diseñado ningún sonido**. Lo que hicimos fue
aprender a leer una fuente real y distinguir:

$$
\text{objeto}
\longrightarrow
\text{mediciones y propiedades}
\longrightarrow
\text{variables candidatas para modelar}
$$

El siguiente paso será preguntar qué variables físicas necesitamos realmente
para AstroWave y cuáles están disponibles en SIMBAD o deberán obtenerse desde
otra fuente.

Esto mantiene nuestro principio rector:

**primero comprender el dato; después decidir cómo representarlo mediante
sonido.**

## 3. ¿Qué variables necesita realmente AstroWave?

Hasta ahora aprendimos a consultar e interpretar algunas propiedades de
Betelgeuse. El siguiente paso no es generar sonido, sino decidir **qué
variables físicas podrían ser relevantes para el proyecto** y comprobar si
podemos obtenerlas de manera reproducible.

Para evitar mezclar preguntas distintas, separaremos tres niveles:

$$
\text{significado físico}
\longrightarrow
\text{disponibilidad del dato}
\longrightarrow
\text{posible representación sonora}
$$

La última columna contiene solamente **hipótesis de trabajo**. No hemos
decidido todavía que esos mappings sean los adecuados.

| Variable candidata | Significado físico | Unidad / tipo | Estado actual | Posible papel acústico |
|---|---|---|---|---|
| Temperatura efectiva | Estado térmico de la estrella | K | Aún no consultada | Pitch o brillo espectral |
| Distancia | Separación respecto del observador | pc, ly | Debemos derivarla u obtenerla de otra fuente | Espacialidad, delay u otro parámetro |
| Luminosidad | Potencia radiada por la estrella | $L_\odot$ u otra unidad | Aún no consultada | Amplitud, densidad u otro parámetro |
| Tipo espectral | Clasificación relacionada con temperatura y propiedades estelares | Categoría | **Disponible:** `sp_type` | Timbre o familia sonora |
| Paralaje | Medición angular relacionada con la distancia | mas | **Disponible:** `plx_value` | Probablemente variable intermedia, no necesariamente sonido directo |
| Tipo de objeto | Clasificación astronómica general | Categoría | **Disponible:** `otype` | Podría utilizarse como atributo categórico |

### Una distinción importante

Que una variable esté disponible **no significa que debamos sonificarla**.

Por ejemplo, la paralaje puede ser muy valiosa para estimar una distancia,
pero quizá sea la **distancia** —y no la paralaje misma— la propiedad que
queramos representar.

De manera similar:

$$
\text{dato disponible}
\neq
\text{variable que queremos comunicar}
\neq
\text{parámetro acústico}
$$

Esta separación nos ayudará a evitar mappings arbitrarios.

## 4. ¿Dónde están las variables que todavía nos faltan?

Nuestra matriz muestra que algunas propiedades ya están disponibles en SIMBAD,
mientras que otras todavía no han sido recuperadas.

Antes de incorporar otra fuente de datos, conviene preguntar qué campos ofrece
realmente SIMBAD.

Nos interesan especialmente tres variables:

- **temperatura efectiva**;
- **luminosidad**;
- **distancia**.

La pregunta no es todavía cuáles son sus valores, sino:

> **¿puede nuestra fuente actual proporcionarlos de manera explícita y
> reproducible?**

Este paso forma parte de la auditoría de los datos.

In [232]:
campos_simbad = Simbad.list_votable_fields()

campos_simbad

name,description,type
object,object,object
mesDiameter,Collection of stellar diameters.,table
mesPM,Collection of proper motions.,table
mesISO,Infrared Space Observatory (ISO) observing log.,table
mesSpT,Collection of spectral types.,table
allfluxes,"all flux/magnitudes U,B,V,I,J,H,K,u_,g_,r_,i_,z_",table
ident,Identifiers of an astronomical object,table
flux,Magnitude/Flux information about an astronomical object,table
mesOtype,Other object types associated with an object with origins,table
mesPLX,Collection of trigonometric parallaxes.,table


### Buscar sin recorrer manualmente 116 opciones

SIMBAD ofrece muchas posibilidades de consulta. En lugar de revisar una por
una, podemos buscar palabras relacionadas con las variables que necesita
AstroWave.

Esta búsqueda no recupera todavía valores de Betelgeuse. Estamos inspeccionando
el **esquema de la fuente de datos**:

$$
\text{variable que necesitamos}
\longrightarrow
\text{¿existe un campo relacionado en SIMBAD?}
$$

Buscaremos inicialmente términos asociados a:

- temperatura;
- luminosidad;
- distancia;
- paralaje.

In [233]:
campos_df = campos_simbad.to_pandas()

terminos = r"temp|temperature|lum|lumin|dist|parallax"

mask = (
    campos_df["name"]
    .astype(str)
    .str.contains(terminos, case=False, regex=True, na=False)
    |
    campos_df["description"]
    .astype(str)
    .str.contains(terminos, case=False, regex=True, na=False)
)

campos_relevantes = (
    campos_df.loc[mask, ["name", "description", "type"]]
    .reset_index(drop=True)
)

campos_relevantes

,name,description,type
0,mesPLX,Collection of trigonometric parallaxes.,table
1,mesDistance,"Collection of distances (pc, kpc or Mpc) by se...",table
2,plx_bibcode,Parallax reference,column of basic
3,plx_err,Parallax error,column of basic
4,plx_err_prec,Parallax error precision,column of basic
5,plx_prec,Parallax precision,column of basic
6,plx_qual,Parallax quality,column of basic
7,plx_value,Parallax,column of basic
8,parallax,all fields related to parallaxes,bundle of basic columns


### Un hallazgo importante: SIMBAD también almacena distancias

Nuestra búsqueda encontró dos tipos de información relacionados con la
distancia:

- `parallax`: conjunto de campos básicos de paralaje;
- `mesPLX`: colección de mediciones de paralaje;
- `mesDistance`: colección de estimaciones de distancia publicadas.

Esto introduce otra distinción importante:

$$
\text{valor resumido o seleccionado}
\neq
\text{colección de mediciones publicadas}
$$

Una misma estrella puede tener varias estimaciones de distancia obtenidas con
métodos o publicaciones diferentes.

Para AstroWave no deberíamos elegir una de ellas automáticamente: primero
debemos conocer qué contiene la colección y qué procedencia tiene cada valor.

In [234]:
distance_columns = Simbad.list_columns("mesDistance")

distance_columns

table_name,column_name,datatype,description,unit,ucd
object,object,object,object,object,object
mesDistance,bibcode,CHAR,measurement bibcode,,meta.bib.bibcode
mesDistance,dist,DOUBLE,Distance value,,pos.distance
mesDistance,dist_prec,SMALLINT,Precision (# of decimal positions) associated with the column dist,,
mesDistance,mespos,SMALLINT,Position of a measurement in a list of measurements,,stat.rank
mesDistance,method,CHAR,distance calculation method,,instr.setup
mesDistance,minus_err,DOUBLE,minus error,,stat.error
mesDistance,minus_err_prec,SMALLINT,Precision (# of decimal positions) associated with the column minus_err,,
mesDistance,oidref,BIGINT,Object internal identifier,,meta.record
mesDistance,plus_err,DOUBLE,plus error,,stat.error


### ¿Qué contiene una medición de distancia?

La colección `mesDistance` no almacena solamente un número. Para cada
estimación puede conservar información como:

- `dist`: valor de distancia;
- `unit`: unidad de la medida (`pc`, `kpc` o `Mpc`);
- `minus_err` y `plus_err`: incertidumbres inferior y superior;
- `method`: método utilizado para estimar la distancia;
- `qual`: indicador de calidad;
- `bibcode`: referencia bibliográfica de la medición.

Esto es importante para AstroWave porque una misma estrella puede tener
**más de una estimación de distancia**.

Por tanto, antes de utilizar una distancia en un modelo de sonificación
debemos conocer su procedencia y decidir qué criterio utilizaremos para
seleccionarla.

In [235]:
simbad_distance = Simbad()

simbad_distance.add_votable_fields("mesDistance")

betelgeuse_distances = simbad_distance.query_object("Betelgeuse")

betelgeuse_distances

main_id,ra,dec,coo_err_maj,coo_err_min,coo_err_angle,coo_wavelength,coo_bibcode,mesdistance.bibcode,mesdistance.dist,mesdistance.dist_prec,mesdistance.mespos,mesdistance.method,mesdistance.minus_err,mesdistance.minus_err_prec,mesdistance.plus_err,mesdistance.plus_err_prec,mesdistance.qual,mesdistance.unit,matched_id
,deg,deg,mas,mas,deg,,,,,,,,,,,,,,
object,float64,float64,float32,float32,int16,str1,object,object,float64,int16,int16,object,float64,int16,float64,int16,str1,object,object
* alf Ori,88.79293899077537,7.407063995272694,9.040423,5.716997,90,O,2007A&A...474..653V,,--,--,--,,--,--,--,--,,,NAME Betelgeuse


### Un resultado también puede ser la ausencia de un dato

La colección `mesDistance` existe en SIMBAD, pero la consulta realizada para
Betelgeuse no devuelve una medición de distancia en esos campos.

Los símbolos `--` indican valores ausentes.

Esto nos permite distinguir tres situaciones:

$$
\text{la variable existe en la fuente}
\neq
\text{la variable está disponible para este objeto}
\neq
\text{tenemos un valor utilizable para el proyecto}
$$

Por tanto, la ausencia de un valor **no es un error del notebook**. Es un
resultado de nuestra auditoría.

Para AstroWave esto significa que, si queremos utilizar distancia, tendremos
que evaluar otra forma reproducible de obtenerla: por ejemplo, derivarla a
partir de una medición apropiada o consultar otra fuente astronómica.

In [236]:
distance_fields = [
    "mesdistance.dist",
    "mesdistance.unit",
    "mesdistance.minus_err",
    "mesdistance.plus_err",
    "mesdistance.method",
    "mesdistance.qual",
    "mesdistance.bibcode",
]

betelgeuse_distances[distance_fields]

mesdistance.dist,mesdistance.unit,mesdistance.minus_err,mesdistance.plus_err,mesdistance.method,mesdistance.qual,mesdistance.bibcode
float64,object,float64,float64,object,str1,object
--,,--,--,,,


In [237]:
for field in distance_fields:
    print(
        f"{field:28s} "
        f"masked = {bool(betelgeuse_distances[field].mask[0])}"
    )

mesdistance.dist             masked = True
mesdistance.unit             masked = False
mesdistance.minus_err        masked = True
mesdistance.plus_err         masked = True
mesdistance.method           masked = False
mesdistance.qual             masked = False
mesdistance.bibcode          masked = False


### Conclusión de esta auditoría

La consulta confirma que `mesDistance` existe como colección en SIMBAD, pero
para Betelgeuse no recuperamos un valor numérico de distancia:

```text
mesdistance.dist → valor ausente
```

Algunos campos asociados aparecen sin un valor científicamente utilizable.
Por eso, para nuestro propósito, la pregunta fundamental no es solamente si
una columna existe, sino:

$$
\text{¿existe un valor físico utilizable para este objeto?}
$$

En esta consulta, para la distancia almacenada en `mesDistance`, la respuesta
es **no**.

Esto no significa que SIMBAD sea una fuente inadecuada. Al contrario, ya nos
ha proporcionado información reproducible sobre:

- identidad del objeto;
- posición;
- tipo de objeto;
- tipo espectral;
- paralaje;
- procedencia bibliográfica de varias mediciones.

La auditoría simplemente muestra que **una misma fuente no tiene por qué
proporcionar todas las variables necesarias para un proyecto**.

Esto nos conduce naturalmente a buscar una fuente complementaria.


## 5. Buscando una fuente complementaria

La auditoría de SIMBAD mostró que una fuente astronómica puede ser muy útil
para algunas propiedades y no proporcionar directamente otras que necesita
AstroWave.

Por ello evaluaremos **fuentes complementarias**, sin reemplazar la información
que ya obtuvimos de SIMBAD.

La estrategia seguirá siendo la misma:

$$
\text{identificar}
\longrightarrow
\text{inspeccionar}
\longrightarrow
\text{auditar}
\longrightarrow
\text{recién después modelar}
$$

Nuestra primera candidata será **Gaia DR3**, porque su catálogo incluye
parámetros astrofísicos potencialmente relevantes para el proyecto, entre
ellos estimaciones de temperatura efectiva, distancia y luminosidad.

Sin embargo, debemos mantener una distinción fundamental:

$$
\text{columna disponible en un catálogo}
\neq
\text{valor disponible para nuestro objeto}
$$

Antes de utilizar cualquier parámetro de Gaia debemos comprobar primero si
podemos identificar de manera fiable a Betelgeuse en ese catálogo.

<details>
<summary><strong>Para profundizar: documentación oficial de Gaia DR3</strong></summary>

- [Gaia DR3: catálogo principal y definición de variables](https://gea.esac.esa.int/archive/documentation/GDR3/Gaia_archive/chap_datamodel/sec_dm_main_source_catalogue/ssec_dm_gaia_source.html)
- [Gaia DR3: contenidos y limitaciones](https://www.cosmos.esa.int/web/gaia/dr3)

</details>


### 5.1. Buscar a Betelgeuse en Gaia a partir de su posición

Ya conocemos la posición de Betelgeuse a partir de SIMBAD.

En lugar de buscar manualmente un identificador de Gaia, utilizaremos esas
coordenadas para realizar una búsqueda espacial alrededor de la estrella.

El procedimiento será:

$$
\text{coordenadas SIMBAD}
\longrightarrow
\text{búsqueda espacial en Gaia DR3}
\longrightarrow
\text{fuentes candidatas}
\longrightarrow
\text{auditoría de la identificación}
$$

Es importante no asumir que la fuente más cercana es automáticamente
Betelgeuse. Primero inspeccionaremos los candidatos y su separación angular.

En este paso **no utilizaremos todavía temperatura, distancia ni luminosidad**.
Primero necesitamos identificar correctamente el objeto.

In [238]:
import astropy.units as u
from astropy.coordinates import SkyCoord
from astroquery.gaia import Gaia

# Coordenadas que ya obtuvimos desde SIMBAD
ra_betelgeuse = float(betelgeuse_extended["ra"][0])
dec_betelgeuse = float(betelgeuse_extended["dec"][0])

betelgeuse_coord = SkyCoord(
    ra=ra_betelgeuse * u.deg,
    dec=dec_betelgeuse * u.deg,
    frame="icrs",
)

print("Coordenadas utilizadas para la búsqueda:")
print(f"RA  = {ra_betelgeuse:.8f} deg")
print(f"Dec = {dec_betelgeuse:.8f} deg")

Coordenadas utilizadas para la búsqueda:
RA  = 88.79293899 deg
Dec = 7.40706400 deg


In [239]:
# Buscaremos fuentes de Gaia DR3 dentro de 60 segundos de arco
# alrededor de la posición proporcionada por SIMBAD.

search_radius = 60 * u.arcsec

Gaia.ROW_LIMIT = 100

job = Gaia.cone_search_async(
    betelgeuse_coord,
    radius=search_radius,
    table_name="gaiadr3.gaia_source",
)

gaia_candidates = job.get_results()

print(f"Número de fuentes candidatas: {len(gaia_candidates)}")

INFO: Query finished. [astroquery.utils.tap.core]
Número de fuentes candidatas: 4


In [240]:
print("Unidad de RA :", gaia_candidates["ra"].unit)
print("Unidad de Dec:", gaia_candidates["dec"].unit)

Unidad de RA : deg
Unidad de Dec: deg


In [241]:
# Gaia ya devuelve ra y dec con unidades angulares.
# No debemos multiplicarlas nuevamente por u.deg.

candidate_coords = SkyCoord(
    ra=gaia_candidates["ra"],
    dec=gaia_candidates["dec"],
    frame="icrs",
)

# Separación angular entre la posición SIMBAD de Betelgeuse
# y cada fuente encontrada en Gaia.
separations = betelgeuse_coord.separation(candidate_coords)

gaia_candidates["separation_arcsec"] = separations.arcsec

# Ordenar desde la fuente angularmente más próxima.
gaia_candidates.sort("separation_arcsec")

candidate_columns = [
    "source_id",
    "ra",
    "dec",
    "separation_arcsec",
    "phot_g_mean_mag",
    "parallax",
    "parallax_error",
    "pmra",
    "pmdec",
    "ruwe",
]

gaia_candidates[candidate_columns][:10]

source_id,ra,dec,separation_arcsec,phot_g_mean_mag,parallax,parallax_error,pmra,pmdec,ruwe
,deg,deg,,mag,mas,mas,mas / yr,mas / yr,
int64,float64,float64,float64,float32,float64,float32,float64,float64,float32
3322962496942145664,88.80291335651158,7.403025485014882,38.46191236678919,13.433729,--,--,--,--,--
3322962565659924352,88.79321019288258,7.420711028184212,49.13885711633251,17.875948,3.9608967019894594,0.17359053,52.26801046904513,9.183620364794656,0.95861685
3322962496943774592,88.80765411285472,7.402206302295387,55.366967373176486,14.895775,0.3955897316051728,0.036055967,0.03988355306406587,-2.0580265479092588,0.968058
3322962668739137152,88.79543948292624,7.423188672896523,58.7311672195862,17.613653,0.7861929181550191,0.1430338,1.9606471613833585,-4.547306732584926,0.98435855


### La fuente más cercana no siempre es el objeto buscado

La búsqueda encontró cuatro fuentes de Gaia DR3 dentro de un radio de
60 segundos de arco alrededor de la posición de Betelgeuse.

Sin embargo, la fuente más próxima se encuentra todavía a aproximadamente
38.5 segundos de arco y las fuentes recuperadas tienen magnitudes \(G\)
entre aproximadamente 13 y 18.

Por tanto, **no tenemos evidencia suficiente para identificar ninguna de estas
fuentes como Betelgeuse**.

Este resultado recuerda una regla importante:

$$
\text{proximidad espacial}
\neq
\text{identidad astronómica}
$$

Gaia presenta además incompletitud en el extremo brillante de su catálogo.
Antes de interpretar la ausencia de una coincidencia cercana, comprobaremos si
SIMBAD conoce algún identificador Gaia asociado explícitamente a Betelgeuse.


In [242]:
betelgeuse_ids = Simbad.query_objectids("Betelgeuse")

betelgeuse_ids

id
object
TIC 269273552
GEN# +1.00039801
HIP 27989
2MASS J05551028+0724255
PLX 1362
* 58 Ori
* alf Ori
AAVSO 0549+07
ADS 4506 AP


### Un mismo objeto puede tener muchos identificadores

SIMBAD devuelve 46 identificadores asociados a Betelgeuse.

Esto ocurre porque distintos catálogos astronómicos pueden asignar nombres o
códigos diferentes al mismo objeto. Por ejemplo, en nuestra consulta aparecen
identificadores procedentes de HIPPARCOS, TIC, 2MASS y otros catálogos.

Por tanto:

$$
\text{objeto físico}
\neq
\text{un único identificador}
$$

Para comprobar si Betelgeuse posee una asociación explícita con Gaia,
filtraremos programáticamente los identificadores que contengan la palabra
`Gaia`, en lugar de revisar manualmente las 46 filas.

In [243]:
ids_df = betelgeuse_ids.to_pandas()

gaia_ids = (
    ids_df[
        ids_df["id"]
        .astype(str)
        .str.contains("Gaia", case=False, na=False)
    ]
    .reset_index(drop=True)
)

gaia_ids

,id


In [244]:
if len(gaia_ids) == 0:
    print("No se encontró ningún identificador Gaia entre los IDs de SIMBAD.")
else:
    print("Identificadores Gaia encontrados:")
    display(gaia_ids)

No se encontró ningún identificador Gaia entre los IDs de SIMBAD.


### Resultado de la identificación en Gaia

SIMBAD devuelve numerosos identificadores para Betelgeuse, pero en esta
consulta **no aparece un identificador Gaia explícito**.

Este resultado, combinado con la búsqueda espacial anterior, indica que no
tenemos una asociación suficientemente sólida como para utilizar una de las
cuatro fuentes cercanas de Gaia como si fuera Betelgeuse.

Por tanto, **no forzaremos un *cross-match* incierto**.

Esto no implica que Gaia sea una fuente inadecuada en general. La documentación
de Gaia señala que el catálogo es incompleto en el extremo brillante, por lo
que objetos excepcionalmente brillantes pueden requerir un tratamiento
especial.

La lista de identificadores de SIMBAD sí contiene una asociación explícita:

```text
TIC 269273552
```

Esto nos ofrece una nueva ruta reproducible hacia el **TESS Input Catalog
(TIC)**.

El aprendizaje metodológico es importante:

$$
\text{una fuente candidata no funciona para este objeto}
\longrightarrow
\text{no forzar la asociación}
\longrightarrow
\text{buscar una ruta con trazabilidad explícita}
$$


### 5.2. TESS Input Catalog: una asociación explícita

SIMBAD identifica a Betelgeuse como `TIC 269273552`.

El **TESS Input Catalog (TIC)** contiene parámetros estelares y permite
consultas mediante identificadores explícitos. Entre sus campos existen
variables potencialmente relevantes para AstroWave, como:

- `Teff`: temperatura efectiva, en K;
- `d`: distancia, en pc;
- `lum`: luminosidad, en unidades solares;
- `plx`: paralaje, en mas;

junto con campos de incertidumbre asociados, como `e_Teff`, `e_d`, `e_lum` y
`e_plx`.

Que esos campos existan en el TIC **no garantiza todavía que Betelgeuse tenga
valores válidos en todos ellos**. Aplicaremos la misma regla que antes:

$$
\text{identificador explícito}
\longrightarrow
\text{consultar el registro}
\longrightarrow
\text{inspeccionar valores y faltantes}
\longrightarrow
\text{auditar su procedencia}
$$

Primero recuperaremos programáticamente el identificador TIC desde la lista de
SIMBAD y luego consultaremos el catálogo. Todavía no seleccionaremos ninguna
variable para la sonificación.



Si el registro contiene una temperatura efectiva utilizable, la emplearemos
más adelante **solo para una demostración sonora mínima**. Esa demostración no
definirá la solución de AstroWave; servirá únicamente para recorrer una vez el
pipeline completo desde un dato real hasta un sonido.

<details>
<summary><strong>Para profundizar: documentación del TIC y astroquery/MAST</strong></summary>

- [MAST: descripción de campos del TESS Input Catalog](https://mast.stsci.edu/api/v0/_t_i_cfields.html)
- [Astroquery: consultas a catálogos MAST](https://astroquery.readthedocs.io/en/stable/mast/mast_catalog.html)

</details>

In [245]:
# Recuperar el identificador TIC directamente desde los IDs de SIMBAD.
tic_matches = (
    ids_df[
        ids_df["id"]
        .astype(str)
        .str.match(r"^TIC\s+\d+$", na=False)
    ]
    .reset_index(drop=True)
)

tic_matches


,id
0,TIC 269273552


In [246]:
from astroquery.mast import Catalogs

if len(tic_matches) != 1:
    raise ValueError(
        f"Esperábamos un único identificador TIC y encontramos {len(tic_matches)}."
    )

tic_id = int(str(tic_matches.loc[0, "id"]).split()[1])

tic_betelgeuse = Catalogs.query_criteria(
    catalog="TIC",
    ID=tic_id,
)

print(f"TIC consultado: {tic_id}")
print(f"Filas recuperadas: {len(tic_betelgeuse)}")

tic_betelgeuse


TIC consultado: 269273552
Filas recuperadas: 1


ID,version,HIP,TYC,UCAC,TWOMASS,SDSS,ALLWISE,GAIA,APASS,KIC,objType,typeSrc,ra,dec,POSflag,pmRA,e_pmRA,pmDEC,e_pmDEC,PMflag,plx,e_plx,PARflag,gallong,gallat,eclong,eclat,Bmag,e_Bmag,Vmag,e_Vmag,umag,e_umag,gmag,e_gmag,rmag,e_rmag,imag,e_imag,zmag,e_zmag,Jmag,e_Jmag,Hmag,e_Hmag,Kmag,e_Kmag,TWOMflag,prox,w1mag,e_w1mag,w2mag,e_w2mag,w3mag,e_w3mag,w4mag,e_w4mag,GAIAmag,e_GAIAmag,Tmag,e_Tmag,TESSflag,SPFlag,Teff,e_Teff,logg,e_logg,MH,e_MH,rad,e_rad,mass,e_mass,rho,e_rho,lumclass,lum,e_lum,d,e_d,ebv,e_ebv,numcont,contratio,disposition,duplicate_id,priority,eneg_EBV,epos_EBV,EBVflag,eneg_Mass,epos_Mass,eneg_Rad,epos_Rad,eneg_rho,epos_rho,eneg_logg,epos_logg,eneg_lum,epos_lum,eneg_dist,epos_dist,distflag,eneg_Teff,epos_Teff,TeffFlag,gaiabp,e_gaiabp,gaiarp,e_gaiarp,gaiaqflag,starchareFlag,VmagFlag,BmagFlag,splists,e_RA,e_Dec,RA_orig,Dec_orig,e_RA_orig,e_Dec_orig,raddflag,wdflag,objID
str9,str8,int64,str1,str10,str16,str1,str19,str1,str1,int64,str4,str5,float64,float64,str5,float64,float64,float64,float64,str1,float64,float64,str1,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,str19,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,str4,str5,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,str1,float64,float64,float64,float64,float64,float64,int64,float64,str1,str1,float64,float64,float64,str1,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,float64,str1,float64,float64,str5,float64,float64,float64,float64,int64,str1,str6,str4,str1,float64,float64,float64,float64,float64,float64,int64,int64,str8
269273552,20190415,--,--,488-015894,05551028+0724255,--,J055510.25+072424.4,--,--,--,STAR,2mass,88.792856,7.407109,2mass,nan,nan,nan,nan,--,nan,nan,--,199.78715,-8.958656,88.754514,-16.026963,0.557,0.121,0.346457,0.023,nan,nan,nan,nan,nan,nan,nan,nan,nan,nan,-2.989,0.103,-4.007,0.162,-4.378,0.186,ACC-333-111-000-0-0,nan,-1.373,0.019,1.747,nan,-2.506,nan,-4.98,0.003,nan,nan,-1.745,0.134,bpjk,spec7,3601.63,41.87,nan,nan,0.0190852,0.086495,nan,nan,nan,nan,nan,nan,--,nan,nan,nan,nan,nan,nan,--,nan,--,--,nan,nan,nan,--,nan,nan,nan,nan,nan,nan,nan,nan,nan,nan,nan,nan,--,nan,nan,spect,nan,nan,nan,nan,--,--,sirful,bpbj,--,310.0,300.0,88.792856,7.407109,310.0,300.0,-1,0,93030011


### 5.3. Leer solo lo necesario del registro TIC

El TIC puede contener muchas columnas. Para este primer recorrido no
necesitamos estudiarlas todas.

Nos concentraremos únicamente en las variables que ya aparecieron en nuestra
matriz de AstroWave y en sus incertidumbres, **si están disponibles**:

```text
temperatura efectiva
distancia
luminosidad
paralaje
```

La siguiente celda construye una vista reducida sin asumir que todos esos
campos existen o contienen valores válidos.


In [247]:
# Vista reducida de las variables candidatas del TIC.
tic_candidate_columns = [
    "ID",
    "Teff",
    "e_Teff",
    "d",
    "e_d",
    "lum",
    "e_lum",
    "plx",
    "e_plx",
]

tic_available_columns = [
    column
    for column in tic_candidate_columns
    if column in tic_betelgeuse.colnames
]

print("Columnas candidatas disponibles:")
print(tic_available_columns)

tic_betelgeuse[tic_available_columns]


Columnas candidatas disponibles:
['ID', 'Teff', 'e_Teff', 'd', 'e_d', 'lum', 'e_lum', 'plx', 'e_plx']


ID,Teff,e_Teff,d,e_d,lum,e_lum,plx,e_plx
str9,float64,float64,float64,float64,float64,float64,float64,float64
269273552,3601.63,41.87,nan,nan,nan,nan,nan,nan


### Qué valores tenemos realmente para cerrar el ejemplo

La consulta al TIC nos permitió comprobar algo importante antes de seguir.

En esta ejecución, Betelgeuse posee una temperatura efectiva utilizable:

$$
T_{\mathrm{eff}} = 3601.63 \pm 41.87\ \mathrm{K}.
$$

En cambio, los campos `d`, `lum` y `plx` del registro TIC aparecen como
valores faltantes (`nan`).

**No rellenaremos esos valores ni los sustituiremos manualmente.**

Para la primera demostración sonora utilizaremos `Teff` del TIC.

Para mostrar brevemente cómo una sonificación puede incorporar una segunda
variable, reutilizaremos una medición que ya auditamos anteriormente en SIMBAD:
la paralaje,

$$
\pi = 6.55 \pm 0.83\ \mathrm{mas}.
$$

No convertiremos esta paralaje en una "distancia definitiva" de Betelgeuse.
La utilizaremos directamente como una variable **relacionada con la distancia**
para construir una segunda hipótesis de representación.


## 6. Una primera sonificación: del dato al sonido

Cerraremos este cuadernillo con **un solo ejemplo audible**.

El objetivo no es definir cómo debe sonar AstroWave, sino recorrer una vez el
pipeline completo con una sola variable:

$$
T
\longrightarrow
g(T)
\longrightarrow
f_{\mathrm{sound}}
\longrightarrow
\text{audio}
$$

Si el registro TIC contiene una temperatura efectiva (`Teff`) utilizable,
la transformaremos en una frecuencia audible.

Para esta demostración elegiremos deliberadamente:

$$
T_{\min}=3000\ \mathrm{K},
\qquad
T_{\max}=10000\ \mathrm{K}
$$

y un intervalo sonoro de dos octavas:

$$
f_{\min}=220\ \mathrm{Hz},
\qquad
f_{\max}=880\ \mathrm{Hz}.
$$

Primero normalizamos la temperatura:

$$
x=
\frac{T-T_{\min}}
     {T_{\max}-T_{\min}},
$$

y luego utilizamos una transformación exponencial:

$$
f_{\mathrm{sound}}
=
f_{\min}
\left(
\frac{f_{\max}}{f_{\min}}
\right)^x.
$$

Los límites de temperatura y frecuencia son **decisiones didácticas para este
ejemplo**. No son propiedades universales de las estrellas ni constituyen el
mapeo definitivo de AstroWave.

Esta función es, nuevamente, una **hipótesis de representación**.


In [248]:
import numpy as np
from IPython.display import Audio, display

# Solo generaremos audio si el TIC contiene una Teff utilizable.
if "Teff" not in tic_betelgeuse.colnames:
    print("El registro TIC no contiene una columna Teff.")
    print("No generaremos audio a partir de una temperatura inexistente.")

else:
    teff_raw = tic_betelgeuse["Teff"][0]

    if np.ma.is_masked(teff_raw):
        print("La temperatura efectiva está ausente en este registro TIC.")
        print("No generaremos audio a partir de un valor faltante.")

    else:
        T = float(teff_raw)

        if not np.isfinite(T):
            print("La temperatura efectiva no contiene un valor finito.")
            print("No generaremos audio a partir de un valor no utilizable.")

        else:
            # Intervalo didáctico de temperatura para esta demostración.
            T_min = 3000.0
            T_max = 10000.0

            # Dos octavas audibles.
            f_min = 220.0
            f_max = 880.0

            # Normalización y limitación al intervalo [0, 1].
            x = np.clip(
                (T - T_min) / (T_max - T_min),
                0.0,
                1.0,
            )

            # Mapping temperatura -> frecuencia.
            f_sound = f_min * (f_max / f_min) ** x

            print(f"Temperatura efectiva:       {T:.1f} K")
            print(f"Valor normalizado:          {x:.4f}")
            print(f"Frecuencia representativa: {f_sound:.1f} Hz")

            # Síntesis de un tono sinusoidal breve.
            sample_rate = 44100
            duration_s = 2.5

            t = np.arange(
                int(sample_rate * duration_s)
            ) / sample_rate

            signal = 0.15 * np.sin(
                2 * np.pi * f_sound * t
            )

            # Fade-in / fade-out de 30 ms para evitar clics.
            fade_samples = int(0.03 * sample_rate)

            envelope = np.ones_like(signal)
            envelope[:fade_samples] = np.linspace(
                0.0, 1.0, fade_samples
            )
            envelope[-fade_samples:] = np.linspace(
                1.0, 0.0, fade_samples
            )

            signal *= envelope

            display(
                Audio(
                    signal,
                    rate=sample_rate,
                    normalize=False,
                )
            )


Temperatura efectiva:       3601.6 K
Valor normalizado:          0.0859
Frecuencia representativa: 247.8 Hz


### 6.1. De una a dos variables: paralaje y reverberación

Hasta ahora utilizamos una sola relación:

$$
T_{\mathrm{eff}}
\longrightarrow
f_{\mathrm{sound}}.
$$

Ahora añadiremos una segunda dimensión perceptual utilizando la paralaje
recuperada desde SIMBAD.

La paralaje es una medida angular relacionada inversamente con la distancia:
en términos generales, **una paralaje menor corresponde a un objeto más
lejano**.

Para esta demostración construiremos:

$$
\pi
\longrightarrow
h(\pi)
\longrightarrow
\text{cantidad de reverberación}.
$$

Por tanto, tendremos dos mappings independientes:

$$
\boxed{
\begin{aligned}
T_{\mathrm{eff}} &\longrightarrow \text{frecuencia},\\
\pi &\longrightarrow \text{reverberación}.
\end{aligned}
}
$$

Usaremos un intervalo didáctico de paralaje entre 1 y 100 mas y una escala
logarítmica. Dentro de ese intervalo, una paralaje menor producirá una mayor
proporción de señal reverberada.

> **Importante:** la reverberación no es una propiedad física de Betelgeuse ni
> representa propagación real de sonido por el espacio. Es una **metáfora
> perceptual** que elegimos para demostrar cómo una segunda variable puede
> controlar otro atributo acústico.

Esta elección es, nuevamente, una **hipótesis de representación**.


In [249]:
# Segunda variable:
# paralaje SIMBAD -> proporción de reverberación.

wet_mix = None

if "plx_value" not in betelgeuse_extended.colnames:
    print("La consulta SIMBAD no contiene `plx_value`.")

else:
    plx_raw = betelgeuse_extended["plx_value"][0]

    if np.ma.is_masked(plx_raw):
        print("La paralaje está ausente en la consulta SIMBAD.")

    else:
        plx_mas = float(plx_raw)

        if not np.isfinite(plx_mas) or plx_mas <= 0:
            print("La paralaje no contiene un valor utilizable.")

        else:
            # Incertidumbre, solo para informar si está disponible.
            plx_err_mas = None
            if "plx_err" in betelgeuse_extended.colnames:
                plx_err_raw = betelgeuse_extended["plx_err"][0]
                if (
                    not np.ma.is_masked(plx_err_raw)
                    and np.isfinite(float(plx_err_raw))
                ):
                    plx_err_mas = float(plx_err_raw)

            # Intervalo didáctico de referencia para la paralaje.
            plx_min = 1.0
            plx_max = 100.0

            # Escala logarítmica:
            # menor paralaje -> mayor proxy perceptual de lejanía.
            plx_clipped = np.clip(
                plx_mas,
                plx_min,
                plx_max,
            )

            distance_proxy = 1.0 - (
                np.log10(plx_clipped) - np.log10(plx_min)
            ) / (
                np.log10(plx_max) - np.log10(plx_min)
            )

            distance_proxy = np.clip(
                distance_proxy,
                0.0,
                1.0,
            )

            # Mantener la reverberación moderada para no ocultar el pitch.
            wet_mix = 0.1 + 0.70 * distance_proxy # 0.10 + 0.40 * distance_proxy

            if plx_err_mas is None:
                print(f"Paralaje SIMBAD:             {plx_mas:.2f} mas")
            else:
                print(
                    "Paralaje SIMBAD:             "
                    f"{plx_mas:.2f} ± {plx_err_mas:.2f} mas"
                )

            print(
                f"Proxy perceptual de lejanía: {distance_proxy:.4f}"
            )
            print(
                f"Mezcla reverberada (wet):    {100 * wet_mix:.1f} %"
            )


Paralaje SIMBAD:             6.55 ± 0.83 mas
Proxy perceptual de lejanía: 0.5919
Mezcla reverberada (wet):    51.4 %


In [250]:
# Comparación A/B:
# A) temperatura -> pitch
# B) temperatura -> pitch + paralaje -> reverberación
#
# La "reverberación" se implementa aquí de forma transparente mediante
# varias reflexiones retardadas y decrecientes. No modela un recinto físico.

if wet_mix is None:
    print("No se generará la comparación porque no hay una paralaje utilizable.")

elif "signal" not in globals():
    print(
        "Primero ejecuta la celda de la sonificación basada en temperatura."
    )

else:
    dry_signal = signal.copy()

    # Reflexiones sintéticas: (retardo en segundos, ganancia relativa)
    #reflections = [
    #    (0.035, 0.55),
    #    (0.070, 0.38),
    #    (0.120, 0.28),
    #    (0.190, 0.20),
    #    (0.290, 0.14),
    #    (0.420, 0.10),
    #]
    reflections = [
        (0.050, 0.60),
        (0.120, 0.45),
        (0.220, 0.35),
        (0.380, 0.25),
        (0.600, 0.18),
        (0.900, 0.12),
    ]

    max_delay_samples = max(
        int(delay_s * sample_rate)
        for delay_s, _ in reflections
    )

    output_length = len(dry_signal) + max_delay_samples

    # Señal seca, extendida para poder comparar con la cola reverberante.
    dry_padded = np.zeros(output_length)
    dry_padded[:len(dry_signal)] = dry_signal

    # Componente reverberada.
    wet_component = np.zeros(output_length)

    for delay_s, gain in reflections:
        delay_samples = int(delay_s * sample_rate)

        wet_component[
            delay_samples:delay_samples + len(dry_signal)
        ] += gain * dry_signal

    # Igualar aproximadamente el pico del componente reverberado
    # al de la señal seca antes de realizar la mezcla.
    dry_peak = np.max(np.abs(dry_signal))
    wet_peak = np.max(np.abs(wet_component))

    if wet_peak > 0:
        wet_component *= dry_peak / wet_peak

    sonified_signal = (
        (1.0 - wet_mix) * dry_padded
        + wet_mix * wet_component
    )

    # Igualar el pico final al de la señal seca para que la comparación
    # se concentre en el efecto y no en una diferencia global de nivel.
    final_peak = np.max(np.abs(sonified_signal))

    if final_peak > 0:
        sonified_signal *= dry_peak / final_peak

    print("A. Temperatura solamente: pitch")
    display(
        Audio(
            dry_padded,
            rate=sample_rate,
            normalize=False,
        )
    )

    print("B. Temperatura + paralaje: pitch + reverberación")
    display(
        Audio(
            sonified_signal,
            rate=sample_rate,
            normalize=False,
        )
    )


A. Temperatura solamente: pitch


B. Temperatura + paralaje: pitch + reverberación


### ¿Qué acabamos de hacer?

Recorrimos primero un pipeline con una sola variable:

$$
T_{\mathrm{eff}}
\longrightarrow
\text{frecuencia}
\longrightarrow
\text{sonido}.
$$

Después construimos una demostración multivariable:

$$
\boxed{
\begin{aligned}
T_{\mathrm{eff}} &\longrightarrow \text{frecuencia},\\
\pi &\longrightarrow \text{reverberación}.
\end{aligned}
}
$$

El sonido que escuchamos **no es Betelgeuse** y la reverberación tampoco es una
propiedad física de la estrella.

Son representaciones construidas a partir de variables astronómicas y reglas
de mapeo que hicimos explícitas.

La comparación A/B permite escuchar una idea central de AstroWave:
**distintas variables pueden controlar dimensiones perceptuales diferentes del
sonido**.

Pero esto abre, más que cerrar, las preguntas verdaderamente interesantes:

- ¿qué variables queremos comunicar?
- ¿qué relaciones entre los datos queremos conservar?
- ¿qué atributo acústico es adecuado para representar cada variable?
- ¿qué mappings producen representaciones perceptualmente útiles?
- ¿cómo podemos comparar distintas alternativas?

No necesitamos resolver esas preguntas ahora. Ese es precisamente el espacio
de exploración que queda abierto para AstroWave.

> **Este cuadernillo no pretende entregar una solución terminada.**
> Su propósito es mostrar una ruta posible: comenzar por datos reales,
> documentar su procedencia, hacer explícitas las decisiones de modelación y
> llegar a un primer sonido reproducible sin perder de vista qué representa y
> de dónde proviene.
